chain/src/linked_list.rs

```diff
@@ -0,0 +1,582 @@
+// Copyright 2020 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implements "linked list" storage primitive for lmdb index supporting multiple entries.
+
+use crate::core::ser::{self, Readable, Reader, Writeable, Writer};
+use crate::store::Batch;
+use crate::types::CommitPos;
+use crate::util::secp::pedersen::Commitment;
+use enum_primitive::FromPrimitive;
+use grin_store as store;
+use std::marker::PhantomData;
+use store::{to_key, to_key_u64, Error};
+
+enum_from_primitive! {
+	#[derive(Copy, Clone, Debug, PartialEq)]
+	enum ListWrapperVariant {
+		Single = 0,
+		Multi = 1,
+	}
+}
+
+impl Writeable for ListWrapperVariant {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
+		writer.write_u8(*self as u8)
+	}
+}
+
+impl Readable for ListWrapperVariant {
+	fn read<R: Reader>(reader: &mut R) -> Result<ListWrapperVariant, ser::Error> {
+		ListWrapperVariant::from_u8(reader.read_u8()?).ok_or(ser::Error::CorruptedData)
+	}
+}
+
+enum_from_primitive! {
+	#[derive(Copy, Clone, Debug, PartialEq)]
+	enum ListEntryVariant {
+		// Start at 2 here to differentiate from ListWrapperVariant above.
+		Head = 2,
+		Tail = 3,
+		Middle = 4,
```

Comment:
Do we really need a linked list with 5 different variants?
Doesn't the database readily support maps from 64-bit ints to 64-bit ints?
We could map the MMR leaf position of an NRD-kernel to the next lower leaf position of an NRD-kernel with the same top k bits, if any. Besides that, we have a 2^k entry map giving the largest MMR leaf position of any NRD-kernel with those top k bits. I guess we could start with k=8 and it would take a long time before NRD kernel usage increases to the point where we'd want to migrate to k=16.